### PR TITLE
Support ID Field for Error and Custom Kafka Topics

### PR DESCRIPTION
**Pull request by mahadeveaswar**
## What does this PR do?

Support ID Field for Error - include consumer type where applicable to read from the kafka topics as needed

## Changes

- [x] Include consumer type in the functions reading from kafka topic
- [x] Add new function to read error_msg topic
- [x] Add kafka input/output topic param in DagDefs

## Before reviewing

- [ ] Are there any linked PRs?
- [ ] Did you test this PR?
- [ ] Is there any migration or seeding necessary?
- [ ] Did you add api to only one of the language (python/typescript)?

### Links:

-   xyme-backend: https://github.com/Accern/xyme-backend/pull/269
-   xyme-notebooks: https://github.com/Accern/xyme-notebooks/pull/83

#### Jira links:
- https://myaccern.atlassian.net/browse/FCN-84
- https://myaccern.atlassian.net/browse/FCN-189

### New API added that not implemented in other languages

If your answer is yes, please also update the todo_api.txt at where the API is missing.

### CODE REVIEW

File: packages/python/accern_xyme/accern_xyme.py [hunk 1]
```diff
@@ -891,16 +891,26 @@ def get_kafka_error_topic(self) -> str:
...
     def delete_kafka_error_topic(self) -> KafkaTopics:
         return cast(KafkaTopics, self.request_json(
             METHOD_POST, "/kafka_topics", {
                 "num_partitions": 0,
             }))
 
-    def read_kafka_errors(self, offset: str = "current") -> List[str]:
+    def read_kafka_errors(
+            self,
+            consumer_type: str,
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:908]: Literal

File: packages/python/accern_xyme/accern_xyme.py [hunk 2]
```diff
@@ -1802,15 +1812,18 @@ def post_kafka_msgs(
 
     def read_kafka_output(
             self,
+            consumer_type: str,
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:1815]: cannot be anything other than output

File: packages/python/accern_xyme/accern_xyme.py (hunk 1 above)
Comment (JosuaKrause) [reply packages/python/accern_xyme/accern_xyme.py:908]: can only be errors

File: packages/typescript/index.ts [hunk 3]
```diff
@@ -1133,12 +1142,16 @@ export default class XYMEClient {
         });
     }
 
-    public async readKafkaErrors(offset: string): Promise<string[]> {
+    public async readKafkaErrors(
+        consumerType: string,
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:1146]: errors only

File: packages/typescript/index.ts [hunk 4]
```diff
@@ -1917,6 +1930,7 @@ export class DagHandle {
     }
 
     public async readKafkaOutput(
+        consumerType: string,
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:1933]: implicitly output

File: packages/python/accern_xyme/accern_xyme.py [hunk 5]
```diff
@@ -891,16 +891,26 @@ def get_kafka_error_topic(self) -> str:
         assert res is not None
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:891]: write function that reads both error topics and matches them up using the id path (note, the function assumes json)

File: packages/python/accern_xyme/accern_xyme.py [hunk 6]
```diff
@@ -891,18 +901,68 @@ def get_kafka_error_topic(self) -> str:
...
+        def parse_input_id_json(json_str: str) -> Optional[str]:
+            try:
+                res = json.loads(json_str)
+                for path in input_id_path:
+                    res = res[path]
+                return res
+            # except json.decoder.JSONDecodeError:
+            #     return None
+            except KeyError:
+                return None
```
Comment (mahadeveaswar) [inline packages/python/accern_xyme/accern_xyme.py:946]: is it just KeyError or do we check for the possibility of json decode error as well and return None?
or Exception in general and return None?

File: packages/python/accern_xyme/accern_xyme.py [hunk 7]
```diff
@@ -174,6 +174,16 @@
 }
 NO_RETRY: List[str] = []  # [METHOD_POST, METHOD_FILE]
 
+ConsumerType = Literal[
+    "dag",
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:178]: dag doesn't need to be a consumer type since you'd only use it implicitly

Comment (JosuaKrause) [reply packages/python/accern_xyme/accern_xyme.py:178]: or maybe create two types

File: packages/python/accern_xyme/accern_xyme.py (hunk 6 above)
Comment (mahadeveaswar) [reply packages/python/accern_xyme/accern_xyme.py:946]: @JosuaKrause Can you check this

File: packages/python/accern_xyme/accern_xyme.py [hunk 8]
```diff
@@ -891,18 +900,69 @@ def get_kafka_error_topic(self) -> str:
...
             METHOD_GET, "/kafka_msg", {
                 "offset": offset,
+                "consumer_type": consumer_type,
             }))
 
+    def read_kafka_full_json_errors(
+            self,
+            input_id_path: List[str]) -> Iterable[Tuple[str, Any]]:
+        errs = self.read_kafka_errors(consumer_type=CONSUMER_ERR)
+        msgs = self.read_kafka_errors(consumer_type=CONSUMER_ERR_MSG)
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:934]: hmm, you wanna do that in a loop maybe? -- the thing is calling read_kafka_full_json_errors multiple times will lose the state so if the err_msg for an err doesn't appear in this call but the next we won't know about the error at the second call anymore

File: packages/python/accern_xyme/accern_xyme.py (hunk 6 above)
Comment (JosuaKrause) [reply packages/python/accern_xyme/accern_xyme.py:946]: yes we ignore json decode errors -- if there is non-json data we skip

File: packages/typescript/index.ts [hunk 9]
```diff
@@ -1133,16 +1147,65 @@ export default class XYMEClient {
...
+                consumer_type: consumerType,
             },
         });
     }
 
+    public async readKafkaFullJsonErrors(
+        inputIdPath: string[]
+    ): Promise<[string, string | undefined][]> {
+        const errs = await this.readKafkaErrors(CONSUMER_ERR);
+        const msgs = await this.readKafkaErrors(CONSUMER_ERR_MSG);
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:1175]: same as the python side

File: packages/typescript/index.ts [hunk 10]
```diff
@@ -1133,16 +1147,65 @@ export default class XYMEClient {
...
+            let jres = JSON.parse(json_str);
+            Array.from(inputIdPath).forEach((path) => {
+                jres = jres[path];
+            });
+            return jres;
+        }
+
+        function parseInputIdText(text: string): string | undefined {
+            const ix = text.search('\ninput_id');
+            if (ix != -1) {
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:1188]: !==

File: packages/typescript/index.ts [hunk 11]
```diff
@@ -1133,16 +1147,65 @@ export default class XYMEClient {
...
+        Array.from(msgs).forEach((msg) => {
+            const inputId = parseInputIdJson(msg);
+            if (inputId != null) {
+                msgLookup.set(inputId, msg);
+            }
+        });
+        const res: [string, string | undefined][] = [];
+        Array.from(errs).forEach((err) => {
+            const inputId = parseInputIdText(err);
+            res.push([err, msgLookup.get(inputId)]);
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:1204]: can we not yield in ts? okay

File: packages/python/accern_xyme/accern_xyme.py [hunk 12]
```diff
@@ -891,18 +900,87 @@ def get_kafka_error_topic(self) -> str:
...
+        Provides information as to what the error is and what is the
+        input id and its associated input message.
+
+        Args:
+            input_id_path (List[str]):
+                The path of the field to be considered as input_id in the
+                input json.
+            msg_lookup (Dict[str, str]):
+                Initially an empty dictionary can be passed for this argument.
+                Consequent calls to this method will require the user to pass
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:944]:
```suggestion
                Subsequent calls to this method will require the user to pass
```

File: packages/typescript/index.ts [hunk 13]
```diff
@@ -1133,16 +1147,76 @@ export default class XYMEClient {
...
         });
     }
 
+    /**
+     * Provides information as to what the error is and what is the
+     * input id and its associated input message.
+     * @param inputIdPath: The path of the field to be considered as
+     * input_id in the input json.
+     * @param msgLookup: Initially an empty dictionary can be passed for
+     * this argument. Consequent calls to this method will require the user
```
Comment (mahadeveaswar) [inline packages/typescript/index.ts:1177]:
```suggestion
     * this argument. Subsequent calls to this method will require the user
```

File: packages/typescript/index.ts [hunk 14]
```diff
@@ -1133,16 +1147,76 @@ export default class XYMEClient {
...
+     * @param msgLookup:
+     * This dictionary will populate with the mappings of input_ids to
+     * messages. For subsequent calls the same dictionary should be used.
+     * Initially an empty dictionary can be passed for this argument.
+     * @returns
+     */
+    public async readKafkaFullJsonErrors(
+        inputIdPath: string[],
+        msgLookup: Map<string, string>
+    ): Promise<[string, string | undefined][]> {
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:1185]: don't return an entry where you cannot find a match -- same for python as well

**Pull request merged**